In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 5GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/flight-fare-prediction-mh/Test_set.xlsx
/kaggle/input/flight-fare-prediction-mh/Sample_submission.xlsx
/kaggle/input/flight-fare-prediction-mh/Data_Train.xlsx


In [5]:
train_df = pd.read_excel('../input/flight-fare-prediction-mh/Data_Train.xlsx', sheet_name=0)
test_df = pd.read_excel('../input/flight-fare-prediction-mh/Test_set.xlsx', sheet_name=0)

In [6]:
train_df.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [16]:
train_df['Data'] = 'Train'
test_df['Data'] = 'Test'
test_df['Price'] = np.nan
test_df = test_df[train_df.columns]

In [17]:
df_all = pd.concat([train_df, test_df], axis=0)

In [18]:
df_all.shape

(13354, 12)

In [19]:
print(train_df.shape)
print(test_df.shape)

(10683, 12)
(2671, 12)


In [20]:
df_all.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 13354 entries, 0 to 2670
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Airline          13354 non-null  object 
 1   Date_of_Journey  13354 non-null  object 
 2   Source           13354 non-null  object 
 3   Destination      13354 non-null  object 
 4   Route            13353 non-null  object 
 5   Dep_Time         13354 non-null  object 
 6   Arrival_Time     13354 non-null  object 
 7   Duration         13354 non-null  object 
 8   Total_Stops      13353 non-null  object 
 9   Additional_Info  13354 non-null  object 
 10  Price            10683 non-null  float64
 11  Data             13354 non-null  object 
dtypes: float64(1), object(11)
memory usage: 1.3+ MB


In [44]:
df_all.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info    Price   Data  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897.0  Train  
1    05:50         13:15   7h 25m     2 stops         No info   7662.0  Train  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882.0  Train  
3    18:05         23:30   5h 25m      1 stop         No info   6218.0  Train  
4    16:50         21:35   4h 45m      1 stop         No info  13302.0  Train

In [23]:
df_all.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 13354 entries, 0 to 2670
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Airline          13354 non-null  object 
 1   Date_of_Journey  13354 non-null  object 
 2   Source           13354 non-null  object 
 3   Destination      13354 non-null  object 
 4   Route            13353 non-null  object 
 5   Dep_Time         13354 non-null  object 
 6   Arrival_Time     13354 non-null  object 
 7   Duration         13354 non-null  object 
 8   Total_Stops      13353 non-null  object 
 9   Additional_Info  13354 non-null  object 
 10  Price            10683 non-null  float64
 11  Data             13354 non-null  object 
dtypes: float64(1), object(11)
memory usage: 1.3+ MB


In [45]:
df_all.Route.dropna(axis=0, inplace=True)

In [46]:
df_all.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 13354 entries, 0 to 2670
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Airline          13354 non-null  object 
 1   Date_of_Journey  13354 non-null  object 
 2   Source           13354 non-null  object 
 3   Destination      13354 non-null  object 
 4   Route            13353 non-null  object 
 5   Dep_Time         13354 non-null  object 
 6   Arrival_Time     13354 non-null  object 
 7   Duration         13354 non-null  object 
 8   Total_Stops      13353 non-null  object 
 9   Additional_Info  13354 non-null  object 
 10  Price            10683 non-null  float64
 11  Data             13354 non-null  object 
dtypes: float64(1), object(11)
memory usage: 1.3+ MB


In [49]:
# As we already have Total_stops column, dropping route column as it is of not much use
df_all.drop('Route', axis=1, inplace=True)

In [50]:
df_all.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 13354 entries, 0 to 2670
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Airline          13354 non-null  object 
 1   Date_of_Journey  13354 non-null  object 
 2   Source           13354 non-null  object 
 3   Destination      13354 non-null  object 
 4   Dep_Time         13354 non-null  object 
 5   Arrival_Time     13354 non-null  object 
 6   Duration         13354 non-null  object 
 7   Total_Stops      13353 non-null  object 
 8   Additional_Info  13354 non-null  object 
 9   Price            10683 non-null  float64
 10  Data             13354 non-null  object 
dtypes: float64(1), object(10)
memory usage: 1.2+ MB


In [51]:
df_all.Total_Stops.value_counts()

1 stop      7056
non-stop    4340
2 stops     1899
3 stops       56
4 stops        2
Name: Total_Stops, dtype: int64

In [52]:
df_all.loc[df_all.Total_Stops.isnull()]

Airline Date_of_Journey Source Destination Dep_Time  Arrival_Time  \
9039  Air India       6/05/2019  Delhi      Cochin    09:45  09:25 07 May   

     Duration Total_Stops Additional_Info   Price   Data  
9039  23h 40m         NaN         No info  7480.0  Train

In [59]:
new_df = df_all.loc[(df_all.Source=='Delhi') & (df_all.Destination=='Cochin') & (df_all.Airline=='Air India') & (df_all.Price<10000)]
new_df

Airline Date_of_Journey Source Destination Dep_Time  Arrival_Time  \
10     Air India       1/06/2019  Delhi      Cochin    09:45         23:00   
61     Air India       3/04/2019  Delhi      Cochin    06:05  09:25 04 Apr   
147    Air India       3/04/2019  Delhi      Cochin    12:30  09:25 04 Apr   
171    Air India      18/04/2019  Delhi      Cochin    18:05         21:10   
191    Air India      27/03/2019  Delhi      Cochin    05:15  06:50 28 Mar   
...          ...             ...    ...         ...      ...           ...   
10573  Air India      27/06/2019  Delhi      Cochin    16:40  19:15 28 Jun   
10578  Air India       9/05/2019  Delhi      Cochin    07:00         19:15   
10652  Air India       9/03/2019  Delhi      Cochin    05:15  06:50 10 Mar   
10661  Air India       9/05/2019  Delhi      Cochin    13:00         19:15   
10669  Air India      15/06/2019  Delhi      Cochin    08:00         19:15   

      Duration Total_Stops Additional_Info   Price   Data  
10     13h 15m      1 stop         No info  8907.0  Train  
61     27h 20m      1 stop         No info  5117.0  Train  
147    20h 55m      1 stop         No info  5117.0  Train  
171      3h 5m    non-stop         No info  8246.0  Train  
191    25h 35m      1 stop         No info  7165.0  Train  
...        ...         ...             ...     ...    ...  
10573  26h 35m     2 stops         No info  9653.0  Train  
10578  12h 15m      1 stop         No info  8669.0  Train  
10652  25h 35m      1 stop         No info  9790.0  Train  
10661   6h 15m      1 stop         No info  8372.0  Train  
10669  11h 15m      1 stop         No info  9929.0  Train  

[398 rows x 11 columns]

In [60]:
new_df.Total_Stops.value_counts()

1 stop      233
2 stops      88
non-stop     75
3 stops       2
Name: Total_Stops, dtype: int64

In [ ]:
# As we see that most of the Air India flights from Delhi to Cochin have 1 stop under price range of 10000, 
# I have replaced Nan value in Total_Stops to 1 Stop

In [57]:
df_all.Total_Stops.fillna('1 stop', inplace=True)

In [61]:
df_all.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 13354 entries, 0 to 2670
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Airline          13354 non-null  object 
 1   Date_of_Journey  13354 non-null  object 
 2   Source           13354 non-null  object 
 3   Destination      13354 non-null  object 
 4   Dep_Time         13354 non-null  object 
 5   Arrival_Time     13354 non-null  object 
 6   Duration         13354 non-null  object 
 7   Total_Stops      13354 non-null  object 
 8   Additional_Info  13354 non-null  object 
 9   Price            10683 non-null  float64
 10  Data             13354 non-null  object 
dtypes: float64(1), object(10)
memory usage: 1.2+ MB


In [62]:
df_all.head()

Airline Date_of_Journey    Source Destination Dep_Time  Arrival_Time  \
0       IndiGo      24/03/2019  Banglore   New Delhi    22:20  01:10 22 Mar   
1    Air India       1/05/2019   Kolkata    Banglore    05:50         13:15   
2  Jet Airways       9/06/2019     Delhi      Cochin    09:25  04:25 10 Jun   
3       IndiGo      12/05/2019   Kolkata    Banglore    18:05         23:30   
4       IndiGo      01/03/2019  Banglore   New Delhi    16:50         21:35   

  Duration Total_Stops Additional_Info    Price   Data  
0   2h 50m    non-stop         No info   3897.0  Train  
1   7h 25m     2 stops         No info   7662.0  Train  
2      19h     2 stops         No info  13882.0  Train  
3   5h 25m      1 stop         No info   6218.0  Train  
4   4h 45m      1 stop         No info  13302.0  Train

In [63]:
df_all['Date_of_Journey'] = pd.to_datetime(df_all['Date_of_Journey'])

In [66]:
df_all['Day_of_Journey'] = df_all.Date_of_Journey.dt.day
df_all['Month_of_Journey'] = df_all.Date_of_Journey.dt.month
df_all['Year_of_Journey'] = df_all.Date_of_Journey.dt.year

In [67]:
df_all.head()

Airline Date_of_Journey    Source Destination Dep_Time  Arrival_Time  \
0       IndiGo      2019-03-24  Banglore   New Delhi    22:20  01:10 22 Mar   
1    Air India      2019-01-05   Kolkata    Banglore    05:50         13:15   
2  Jet Airways      2019-09-06     Delhi      Cochin    09:25  04:25 10 Jun   
3       IndiGo      2019-12-05   Kolkata    Banglore    18:05         23:30   
4       IndiGo      2019-01-03  Banglore   New Delhi    16:50         21:35   

  Duration Total_Stops Additional_Info    Price   Data  Day_of_Journey  \
0   2h 50m    non-stop         No info   3897.0  Train              24   
1   7h 25m     2 stops         No info   7662.0  Train               5   
2      19h     2 stops         No info  13882.0  Train               6   
3   5h 25m      1 stop         No info   6218.0  Train               5   
4   4h 45m      1 stop         No info  13302.0  Train               3   

   Month_of_Journey  Year_of_Journey  
0                 3             2019  
1                 1             2019  
2                 9             2019  
3                12             2019  
4                 1             2019

In [68]:
df_all.drop('Date_of_Journey', axis=1, inplace=True)

In [69]:
df_all.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 13354 entries, 0 to 2670
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Airline           13354 non-null  object 
 1   Source            13354 non-null  object 
 2   Destination       13354 non-null  object 
 3   Dep_Time          13354 non-null  object 
 4   Arrival_Time      13354 non-null  object 
 5   Duration          13354 non-null  object 
 6   Total_Stops       13354 non-null  object 
 7   Additional_Info   13354 non-null  object 
 8   Price             10683 non-null  float64
 9   Data              13354 non-null  object 
 10  Day_of_Journey    13354 non-null  int64  
 11  Month_of_Journey  13354 non-null  int64  
 12  Year_of_Journey   13354 non-null  int64  
dtypes: float64(1), int64(3), object(9)
memory usage: 1.4+ MB


In [81]:
time = '23:30'
time
print(type(time))

<class 'str'>


In [78]:
time.split()[0]

'23:30'

In [84]:
df_all.Arrival_Time = df_all.Arrival_Time.str.split(expand=True)[0]

In [85]:
df_all.head()

Airline    Source Destination Dep_Time Arrival_Time Duration  \
0       IndiGo  Banglore   New Delhi    22:20        01:10   2h 50m   
1    Air India   Kolkata    Banglore    05:50        13:15   7h 25m   
2  Jet Airways     Delhi      Cochin    09:25        04:25      19h   
3       IndiGo   Kolkata    Banglore    18:05        23:30   5h 25m   
4       IndiGo  Banglore   New Delhi    16:50        21:35   4h 45m   

  Total_Stops Additional_Info    Price   Data  Day_of_Journey  \
0    non-stop         No info   3897.0  Train              24   
1     2 stops         No info   7662.0  Train               5   
2     2 stops         No info  13882.0  Train               6   
3      1 stop         No info   6218.0  Train               5   
4      1 stop         No info  13302.0  Train               3   

   Month_of_Journey  Year_of_Journey  
0                 3             2019  
1                 1             2019  
2                 9             2019  
3                12             2019  
4                 1             2019

In [87]:
df_all['Duration_hrs'] = df_all.Duration.str.split(expand=True)[0]
df_all['Duration_mins'] = df_all.Duration.str.split(expand=True)[1]

In [88]:
df_all.head()

Airline    Source Destination Dep_Time Arrival_Time Duration  \
0       IndiGo  Banglore   New Delhi    22:20        01:10   2h 50m   
1    Air India   Kolkata    Banglore    05:50        13:15   7h 25m   
2  Jet Airways     Delhi      Cochin    09:25        04:25      19h   
3       IndiGo   Kolkata    Banglore    18:05        23:30   5h 25m   
4       IndiGo  Banglore   New Delhi    16:50        21:35   4h 45m   

  Total_Stops Additional_Info    Price   Data  Day_of_Journey  \
0    non-stop         No info   3897.0  Train              24   
1     2 stops         No info   7662.0  Train               5   
2     2 stops         No info  13882.0  Train               6   
3      1 stop         No info   6218.0  Train               5   
4      1 stop         No info  13302.0  Train               3   

   Month_of_Journey  Year_of_Journey Duration_hrs Duration_mins  
0                 3             2019           2h           50m  
1                 1             2019           7h           25m  
2                 9             2019          19h          None  
3                12             2019           5h           25m  
4                 1             2019           4h           45m

In [89]:
# Time taken by plane to reach destination is called Duration
# It is the differnce betwwen Departure Time and Arrival time


# Assigning and converting Duration column into list
duration = list(df_all["Duration"])

for i in range(len(duration)):
    if len(duration[i].split()) != 2:    # Check if duration contains only hour or mins
        if "h" in duration[i]:
            duration[i] = duration[i].strip() + " 0m"   # Adds 0 minute
        else:
            duration[i] = "0h " + duration[i]           # Adds 0 hour

duration_hrs = []
duration_mins = []
for i in range(len(duration)):
    duration_hrs.append(int(duration[i].split(sep = "h")[0]))    # Extract hours from duration
    duration_mins.append(int(duration[i].split(sep = "m")[0].split()[-1]))   # Extracts only minutes from duration

In [90]:
df_all.Duration_hrs = duration_hrs
df_all.Duration_mins = duration_mins

In [91]:
df_all.head()

Airline    Source Destination Dep_Time Arrival_Time Duration  \
0       IndiGo  Banglore   New Delhi    22:20        01:10   2h 50m   
1    Air India   Kolkata    Banglore    05:50        13:15   7h 25m   
2  Jet Airways     Delhi      Cochin    09:25        04:25      19h   
3       IndiGo   Kolkata    Banglore    18:05        23:30   5h 25m   
4       IndiGo  Banglore   New Delhi    16:50        21:35   4h 45m   

  Total_Stops Additional_Info    Price   Data  Day_of_Journey  \
0    non-stop         No info   3897.0  Train              24   
1     2 stops         No info   7662.0  Train               5   
2     2 stops         No info  13882.0  Train               6   
3      1 stop         No info   6218.0  Train               5   
4      1 stop         No info  13302.0  Train               3   

   Month_of_Journey  Year_of_Journey  Duration_hrs  Duration_mins  
0                 3             2019             2             50  
1                 1             2019             7             25  
2                 9             2019            19              0  
3                12             2019             5             25  
4                 1             2019             4             45

In [92]:
df_all.drop('Duration', axis=1, inplace=True)

In [93]:
df_all['Dep_hr'] = pd.to_datetime(df_all.Dep_Time).dt.hour
df_all['Dep_min'] = pd.to_datetime(df_all.Dep_Time).dt.minute
df_all.drop('Dep_Time', axis=1, inplace=True)

In [94]:
df_all['Arrival_hr'] = pd.to_datetime(df_all.Arrival_Time).dt.hour
df_all['Arrival_min'] = pd.to_datetime(df_all.Arrival_Time).dt.minute
df_all.drop('Arrival_Time', axis=1, inplace=True)

In [95]:
df_all.head()

Airline    Source Destination Total_Stops Additional_Info    Price  \
0       IndiGo  Banglore   New Delhi    non-stop         No info   3897.0   
1    Air India   Kolkata    Banglore     2 stops         No info   7662.0   
2  Jet Airways     Delhi      Cochin     2 stops         No info  13882.0   
3       IndiGo   Kolkata    Banglore      1 stop         No info   6218.0   
4       IndiGo  Banglore   New Delhi      1 stop         No info  13302.0   

    Data  Day_of_Journey  Month_of_Journey  Year_of_Journey  Duration_hrs  \
0  Train              24                 3             2019             2   
1  Train               5                 1             2019             7   
2  Train               6                 9             2019            19   
3  Train               5                12             2019             5   
4  Train               3                 1             2019             4   

   Duration_mins  Dep_hr  Dep_min  Arrival_hr  Arrival_min  
0             50      22       20           1           10  
1             25       5       50          13           15  
2              0       9       25           4           25  
3             25      18        5          23           30  
4             45      16       50          21           35

In [96]:
df_all.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 13354 entries, 0 to 2670
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Airline           13354 non-null  object 
 1   Source            13354 non-null  object 
 2   Destination       13354 non-null  object 
 3   Total_Stops       13354 non-null  object 
 4   Additional_Info   13354 non-null  object 
 5   Price             10683 non-null  float64
 6   Data              13354 non-null  object 
 7   Day_of_Journey    13354 non-null  int64  
 8   Month_of_Journey  13354 non-null  int64  
 9   Year_of_Journey   13354 non-null  int64  
 10  Duration_hrs      13354 non-null  int64  
 11  Duration_mins     13354 non-null  int64  
 12  Dep_hr            13354 non-null  int64  
 13  Dep_min           13354 non-null  int64  
 14  Arrival_hr        13354 non-null  int64  
 15  Arrival_min       13354 non-null  int64  
dtypes: float64(1), int64(9), object(6)
memory

In [97]:
df_all.select_dtypes('object').columns

Index(['Airline', 'Source', 'Destination', 'Total_Stops', 'Additional_Info',
       'Data'],
      dtype='object')

In [100]:
df_all_dummy = pd.get_dummies(data=df_all, columns=['Airline', 'Source', 'Destination', 'Total_Stops', 'Additional_Info'], drop_first=True)

In [101]:
df_all_dummy.head()

Price   Data  Day_of_Journey  Month_of_Journey  Year_of_Journey  \
0   3897.0  Train              24                 3             2019   
1   7662.0  Train               5                 1             2019   
2  13882.0  Train               6                 9             2019   
3   6218.0  Train               5                12             2019   
4  13302.0  Train               3                 1             2019   

   Duration_hrs  Duration_mins  Dep_hr  Dep_min  Arrival_hr  ...  \
0             2             50      22       20           1  ...   
1             7             25       5       50          13  ...   
2            19              0       9       25           4  ...   
3             5             25      18        5          23  ...   
4             4             45      16       50          21  ...   

   Total_Stops_non-stop  Additional_Info_1 Short layover  \
0                     1                                0   
1                     0                                0   
2                     0                                0   
3                     0                                0   
4                     0                                0   

   Additional_Info_2 Long layover  Additional_Info_Business class  \
0                               0                               0   
1                               0                               0   
2                               0                               0   
3                               0                               0   
4                               0                               0   

   Additional_Info_Change airports  \
0                                0   
1                                0   
2                                0   
3                                0   
4                                0   

   Additional_Info_In-flight meal not included  Additional_Info_No Info  \
0                                            0                        0   
1                                            0                        0   
2                                            0                        0   
3                                            0                        0   
4                                            0                        0   

   Additional_Info_No check-in baggage included  Additional_Info_No info  \
0                                             0                        1   
1                                             0                        1   
2                                             0                        1   
3                                             0                        1   
4                                             0                        1   

   Additional_Info_Red-eye flight  
0                               0  
1                               0  
2                               0  
3                               0  
4                               0  

[5 rows x 44 columns]

In [102]:
train_set = df_all_dummy[df_all_dummy.Data=='Train']
test_set = df_all_dummy[df_all_dummy.Data=='Test']

In [103]:
train_set.drop('Data', axis=1, inplace=True)
test_set.drop(['Data','Price'], axis=1, inplace=True)

/opt/conda/lib/python3.7/site-packages/pandas/core/frame.py:3997: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [104]:
print(train_set.shape)
print(test_set.shape)

(10683, 43)
(2671, 42)


In [107]:
X = train_set.drop('Price', axis=1)
y = train_set['Price']

In [108]:
from sklearn.ensemble import ExtraTreesRegressor
etreg = ExtraTreesRegressor(n_estimators=500, max_features=10, random_state=2)
etreg.fit(X, y)

ExtraTreesRegressor(max_features=10, n_estimators=500, random_state=2)

In [114]:
features_imp = pd.DataFrame(etreg.feature_importances_, index=X.columns, columns=['Score'])

In [117]:
features_imp.nlargest(20, columns=['Score']).index

Index(['Total_Stops_non-stop', 'Airline_Jet Airways', 'Day_of_Journey',
       'Month_of_Journey', 'Duration_hrs', 'Airline_IndiGo',
       'Airline_Jet Airways Business',
       'Additional_Info_In-flight meal not included', 'Total_Stops_2 stops',
       'Additional_Info_No info', 'Destination_New Delhi', 'Arrival_hr',
       'Dep_hr', 'Dep_min', 'Duration_mins', 'Arrival_min', 'Airline_SpiceJet',
       'Destination_Delhi', 'Additional_Info_Business class',
       'Airline_Multiple carriers'],
      dtype='object')

In [118]:
X_imp_feat = X[['Total_Stops_non-stop', 'Airline_Jet Airways', 'Day_of_Journey',
       'Month_of_Journey', 'Duration_hrs', 'Airline_IndiGo',
       'Airline_Jet Airways Business',
       'Additional_Info_In-flight meal not included', 'Total_Stops_2 stops',
       'Additional_Info_No info', 'Destination_New Delhi', 'Arrival_hr',
       'Dep_hr', 'Dep_min', 'Duration_mins', 'Arrival_min', 'Airline_SpiceJet',
       'Destination_Delhi', 'Additional_Info_Business class',
       'Airline_Multiple carriers']]

In [119]:
X_imp_feat

Total_Stops_non-stop  Airline_Jet Airways  Day_of_Journey  \
0                         1                    0              24   
1                         0                    0               5   
2                         0                    1               6   
3                         0                    0               5   
4                         0                    0               3   
...                     ...                  ...             ...   
10678                     1                    0               4   
10679                     1                    0              27   
10680                     1                    1              27   
10681                     1                    0               3   
10682                     0                    0               5   

       Month_of_Journey  Duration_hrs  Airline_IndiGo  \
0                     3             2               1   
1                     1             7               0   
2                     9            19               0   
3                    12             5               1   
4                     1             4               1   
...                 ...           ...             ...   
10678                 9             2               0   
10679                 4             2               0   
10680                 4             3               0   
10681                 1             2               0   
10682                 9             8               0   

       Airline_Jet Airways Business  \
0                                 0   
1                                 0   
2                                 0   
3                                 0   
4                                 0   
...                             ...   
10678                             0   
10679                             0   
10680                             0   
10681                             0   
10682                             0   

       Additional_Info_In-flight meal not included  Total_Stops_2 stops  \
0                                                0                    0   
1                                                0                    1   
2                                                0                    1   
3                                                0                    0   
4                                                0                    0   
...                                            ...                  ...   
10678                                            0                    0   
10679                                            0                    0   
10680                                            0                    0   
10681                                            0                    0   
10682                                            0                    1   

       Additional_Info_No info  Destination_New Delhi  Arrival_hr  Dep_hr  \
0                            1                      1           1      22   
1                            1                      0          13       5   
2                            1                      0           4       9   
3                            1                      0          23      18   
4                            1                      1          21      16   
...                        ...                    ...         ...     ...   
10678                        1                      0          22      19   
10679                        1                      0          23      20   
10680                        1                      0          11       8   
10681                        1                      1          14      11   
10682                        1                      0          19      10   

       Dep_min  Duration_mins  Arrival_min  Airline_SpiceJet  \
0           20             50           10                 0   
1           50             25           15                 0   
2           25    

In [120]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_imp_feat, y, test_size = 0.2, random_state=2)

In [121]:
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(8546, 20)
(2137, 20)
(8546,)
(2137,)


In [122]:
from sklearn.linear_model import LinearRegression, Ridge, Lasso

In [123]:
regressor = LinearRegression()

In [124]:
regressor.fit(X_train, y_train)

LinearRegression()

In [125]:
y_hat = regressor.predict(X_test)
y_hat

array([4066.77680785, 7937.48499761, 4709.83509344, ..., 9536.39205369,
       3705.53876804, 9313.87685209])

In [126]:
from sklearn.metrics import mean_squared_error
import math

In [127]:
rmse = math.sqrt(mean_squared_error(y_test, y_hat))
rmse

2600.791991128351

In [133]:
param = {'alpha' : [0.00001, 0.0001, 0.001, 0.01, 0.1, 0, 1, 10, 100, 1000]}
from sklearn.model_selection import GridSearchCV
regressor1 = Lasso()
cross_val = GridSearchCV(regressor1, param_grid=param, cv=10, verbose=3)
cross_val.fit(X_train, y_train)

Fitting 10 folds for each of 10 candidates, totalling 100 fits
[CV] alpha=1e-05 .....................................................
[CV] ......................... alpha=1e-05, score=0.651, total=   0.0s
[CV] alpha=1e-05 .....................................................
[CV] ......................... alpha=1e-05, score=0.623, total=   0.0s
[CV] alpha=1e-05 .....................................................
[CV] ......................... alpha=1e-05, score=0.746, total=   0.0s
[CV] alpha=1e-05 .....................................................
[CV] ......................... alpha=1e-05, score=0.705, total=   0.0s
[CV] alpha=1e-05 .....................................................
[CV] ......................... alpha=1e-05, score=0.687, total=   0.0s
[CV] alpha=1e-05 .....................................................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    0.1s remaining:    0.0s


[CV] ......................... alpha=1e-05, score=0.637, total=   0.0s
[CV] alpha=1e-05 .....................................................
[CV] ......................... alpha=1e-05, score=0.649, total=   0.0s
[CV] alpha=1e-05 .....................................................
[CV] ......................... alpha=1e-05, score=0.658, total=   0.0s
[CV] alpha=1e-05 .....................................................
[CV] ......................... alpha=1e-05, score=0.670, total=   0.0s
[CV] alpha=1e-05 .....................................................
[CV] ......................... alpha=1e-05, score=0.673, total=   0.0s
[CV] alpha=0.0001 ....................................................
[CV] ........................ alpha=0.0001, score=0.651, total=   0.0s
[CV] alpha=0.0001 ....................................................
[CV] ........................ alpha=0.0001, score=0.623, total=   0.0s
[CV] alpha=0.0001 ....................................................
[CV] .

/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:531: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  estimator.fit(X_train, y_train, **fit_params)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: UserWarning: Coordinate descent with no regularization may lead to unexpected results and is discouraged.
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 27070230839.423607, tolerance: 16777743.261533637
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:531: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  estimator.fit(X_train, y_train, **fit_params)
/opt/conda/lib/python3.7/sit

[CV] ............................. alpha=0, score=0.651, total=   0.2s
[CV] alpha=0 .........................................................
[CV] ............................. alpha=0, score=0.623, total=   0.2s
[CV] alpha=0 .........................................................


/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 26051577648.568115, tolerance: 16357256.123308878
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:531: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  estimator.fit(X_train, y_train, **fit_params)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: UserWarning: Coordinate descent with no regularization may lead to unexpected results and is discouraged.
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 27051575857.193794, tolerance: 16083656.283278247
  positive)
/opt/conda/lib/

[CV] ............................. alpha=0, score=0.746, total=   0.2s
[CV] alpha=0 .........................................................
[CV] ............................. alpha=0, score=0.705, total=   0.2s
[CV] alpha=0 .........................................................


/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 27479809903.888054, tolerance: 16743799.20841061
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:531: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  estimator.fit(X_train, y_train, **fit_params)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: UserWarning: Coordinate descent with no regularization may lead to unexpected results and is discouraged.
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 27452807472.5769, tolerance: 16829330.89104369
  positive)
/opt/conda/lib/pyth

[CV] ............................. alpha=0, score=0.687, total=   0.2s
[CV] alpha=0 .........................................................
[CV] ............................. alpha=0, score=0.637, total=   0.2s
[CV] alpha=0 .........................................................


/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 26582513896.205715, tolerance: 16567316.511229541
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:531: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  estimator.fit(X_train, y_train, **fit_params)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: UserWarning: Coordinate descent with no regularization may lead to unexpected results and is discouraged.
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 26220359778.1074, tolerance: 16262051.506873365
  positive)
/opt/conda/lib/py

[CV] ............................. alpha=0, score=0.649, total=   0.2s
[CV] alpha=0 .........................................................
[CV] ............................. alpha=0, score=0.658, total=   0.2s
[CV] alpha=0 .........................................................


/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 27177286067.09194, tolerance: 16806198.291303106
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:531: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  estimator.fit(X_train, y_train, **fit_params)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: UserWarning: Coordinate descent with no regularization may lead to unexpected results and is discouraged.
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 27405130613.68491, tolerance: 16881383.096451208
  positive)
/opt/conda/lib/py

[CV] ............................. alpha=0, score=0.670, total=   0.2s
[CV] alpha=0 .........................................................
[CV] ............................. alpha=0, score=0.673, total=   0.2s
[CV] alpha=1 .........................................................
[CV] ............................. alpha=1, score=0.651, total=   0.0s
[CV] alpha=1 .........................................................


/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 27022991717.927788, tolerance: 16630922.018803846
  positive)


[CV] ............................. alpha=1, score=0.623, total=   0.0s
[CV] alpha=1 .........................................................
[CV] ............................. alpha=1, score=0.735, total=   0.0s
[CV] alpha=1 .........................................................
[CV] ............................. alpha=1, score=0.706, total=   0.0s
[CV] alpha=1 .........................................................
[CV] ............................. alpha=1, score=0.687, total=   0.0s
[CV] alpha=1 .........................................................
[CV] ............................. alpha=1, score=0.637, total=   0.0s
[CV] alpha=1 .........................................................
[CV] ............................. alpha=1, score=0.656, total=   0.0s
[CV] alpha=1 .........................................................
[CV] ............................. alpha=1, score=0.658, total=   0.0s
[CV] alpha=1 .........................................................
[CV] .

[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    3.6s finished
/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_search.py:765: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  self.best_estimator_.fit(X, y, **fit_params)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: UserWarning: Coordinate descent with no regularization may lead to unexpected results and is discouraged.
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 29960943986.65363, tolerance: 18437980.649055984
  positive)


GridSearchCV(cv=10, estimator=Lasso(),
             param_grid={'alpha': [1e-05, 0.0001, 0.001, 0.01, 0.1, 0, 1, 10,
                                   100, 1000]},
             verbose=3)

In [134]:
cross_val.best_estimator_

Lasso(alpha=0)

In [135]:
regressor1 = Lasso(alpha=0)
regressor1.fit(X_train, y_train)
y_hat1 = regressor1.predict(X_test)
rmse = math.sqrt(mean_squared_error(y_test, y_hat1))
rmse

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:2: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: UserWarning: Coordinate descent with no regularization may lead to unexpected results and is discouraged.
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 29960943986.65363, tolerance: 18437980.649055984
  positive)


2600.7919911283543

In [136]:
regressor2 = Ridge()
cross_val1 = GridSearchCV(regressor1, param_grid=param, cv=10, verbose=3)
cross_val1.fit(X_train, y_train)

Fitting 10 folds for each of 10 candidates, totalling 100 fits
[CV] alpha=1e-05 .....................................................
[CV] ......................... alpha=1e-05, score=0.651, total=   0.0s
[CV] alpha=1e-05 .....................................................
[CV] ......................... alpha=1e-05, score=0.623, total=   0.0s
[CV] alpha=1e-05 .....................................................
[CV] ......................... alpha=1e-05, score=0.746, total=   0.0s
[CV] alpha=1e-05 .....................................................
[CV] ......................... alpha=1e-05, score=0.705, total=   0.0s
[CV] alpha=1e-05 .....................................................
[CV] ......................... alpha=1e-05, score=0.687, total=   0.0s
[CV] alpha=1e-05 .....................................................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    0.1s remaining:    0.0s


[CV] ......................... alpha=1e-05, score=0.637, total=   0.0s
[CV] alpha=1e-05 .....................................................
[CV] ......................... alpha=1e-05, score=0.649, total=   0.0s
[CV] alpha=1e-05 .....................................................
[CV] ......................... alpha=1e-05, score=0.658, total=   0.0s
[CV] alpha=1e-05 .....................................................
[CV] ......................... alpha=1e-05, score=0.670, total=   0.0s
[CV] alpha=1e-05 .....................................................
[CV] ......................... alpha=1e-05, score=0.673, total=   0.0s
[CV] alpha=0.0001 ....................................................
[CV] ........................ alpha=0.0001, score=0.651, total=   0.0s
[CV] alpha=0.0001 ....................................................
[CV] ........................ alpha=0.0001, score=0.623, total=   0.0s
[CV] alpha=0.0001 ....................................................
[CV] .

/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:531: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  estimator.fit(X_train, y_train, **fit_params)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: UserWarning: Coordinate descent with no regularization may lead to unexpected results and is discouraged.
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 27070230839.423607, tolerance: 16777743.261533637
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:531: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  estimator.fit(X_train, y_train, **fit_params)
/opt/conda/lib/python3.7/sit

[CV] ............................. alpha=0, score=0.651, total=   0.2s
[CV] alpha=0 .........................................................
[CV] ............................. alpha=0, score=0.623, total=   0.2s
[CV] alpha=0 .........................................................


/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 26051577648.568115, tolerance: 16357256.123308878
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:531: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  estimator.fit(X_train, y_train, **fit_params)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: UserWarning: Coordinate descent with no regularization may lead to unexpected results and is discouraged.
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 27051575857.193794, tolerance: 16083656.283278247
  positive)
/opt/conda/lib/

[CV] ............................. alpha=0, score=0.746, total=   0.2s
[CV] alpha=0 .........................................................
[CV] ............................. alpha=0, score=0.705, total=   0.2s
[CV] alpha=0 .........................................................


/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 27479809903.888054, tolerance: 16743799.20841061
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:531: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  estimator.fit(X_train, y_train, **fit_params)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: UserWarning: Coordinate descent with no regularization may lead to unexpected results and is discouraged.
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 27452807472.5769, tolerance: 16829330.89104369
  positive)
/opt/conda/lib/pyth

[CV] ............................. alpha=0, score=0.687, total=   0.2s
[CV] alpha=0 .........................................................
[CV] ............................. alpha=0, score=0.637, total=   0.2s
[CV] alpha=0 .........................................................


/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 26582513896.205715, tolerance: 16567316.511229541
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:531: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  estimator.fit(X_train, y_train, **fit_params)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: UserWarning: Coordinate descent with no regularization may lead to unexpected results and is discouraged.
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 26220359778.1074, tolerance: 16262051.506873365
  positive)
/opt/conda/lib/py

[CV] ............................. alpha=0, score=0.649, total=   0.2s
[CV] alpha=0 .........................................................
[CV] ............................. alpha=0, score=0.658, total=   0.2s
[CV] alpha=0 .........................................................


/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 27177286067.09194, tolerance: 16806198.291303106
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:531: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  estimator.fit(X_train, y_train, **fit_params)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: UserWarning: Coordinate descent with no regularization may lead to unexpected results and is discouraged.
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 27405130613.68491, tolerance: 16881383.096451208
  positive)
/opt/conda/lib/py

[CV] ............................. alpha=0, score=0.670, total=   0.2s
[CV] alpha=0 .........................................................
[CV] ............................. alpha=0, score=0.673, total=   0.2s
[CV] alpha=1 .........................................................
[CV] ............................. alpha=1, score=0.651, total=   0.0s
[CV] alpha=1 .........................................................


/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 27022991717.927788, tolerance: 16630922.018803846
  positive)


[CV] ............................. alpha=1, score=0.623, total=   0.0s
[CV] alpha=1 .........................................................
[CV] ............................. alpha=1, score=0.735, total=   0.0s
[CV] alpha=1 .........................................................
[CV] ............................. alpha=1, score=0.706, total=   0.0s
[CV] alpha=1 .........................................................
[CV] ............................. alpha=1, score=0.687, total=   0.0s
[CV] alpha=1 .........................................................
[CV] ............................. alpha=1, score=0.637, total=   0.0s
[CV] alpha=1 .........................................................
[CV] ............................. alpha=1, score=0.656, total=   0.0s
[CV] alpha=1 .........................................................
[CV] ............................. alpha=1, score=0.658, total=   0.0s
[CV] alpha=1 .........................................................
[CV] .

[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    3.6s finished
/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_search.py:765: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  self.best_estimator_.fit(X, y, **fit_params)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: UserWarning: Coordinate descent with no regularization may lead to unexpected results and is discouraged.
  positive)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 29960943986.65363, tolerance: 18437980.649055984
  positive)


GridSearchCV(cv=10, estimator=Lasso(alpha=0),
             param_grid={'alpha': [1e-05, 0.0001, 0.001, 0.01, 0.1, 0, 1, 10,
                                   100, 1000]},
             verbose=3)

In [138]:
cross_val1.best_params_

{'alpha': 0}

In [139]:
regressor1 = Ridge(alpha=0)
regressor2.fit(X_train, y_train)
y_hat2 = regressor2.predict(X_test)
rmse = math.sqrt(mean_squared_error(y_test, y_hat2))
rmse

2600.210694257697

In [141]:
from sklearn.ensemble import RandomForestRegressor
from xgboost.sklearn import XGBRegressor
from sklearn.model_selection import RandomizedSearchCV

In [142]:
params = {'n_estimators' : [100, 300, 500, 700, 900, 1100, 1300, 1500],
         'max_features' : [5, 7, 10], 
         'max_depth' : [3, 4, 5, 6, 7], 
         'min_samples_split' : [10, 15, 20], 
         'min_samples_leaf' : [10, 15, 20]}

In [143]:
rf_reg = RandomForestRegressor(random_state=2)

In [144]:
random_cv = RandomizedSearchCV(rf_reg, param_distributions=params, cv=10, verbose=1)

In [145]:
random_cv.fit(X_train, y_train)

Fitting 10 folds for each of 10 candidates, totalling 100 fits


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:  6.4min finished


RandomizedSearchCV(cv=10, estimator=RandomForestRegressor(random_state=2),
                   param_distributions={'max_depth': [3, 4, 5, 6, 7],
                                        'max_features': [5, 7, 10],
                                        'min_samples_leaf': [10, 15, 20],
                                        'min_samples_split': [10, 15, 20],
                                        'n_estimators': [100, 300, 500, 700,
                                                         900, 1100, 1300,
                                                         1500]},
                   verbose=1)

In [150]:
random_cv.best_estimator_

RandomForestRegressor(max_depth=7, max_features=10, min_samples_leaf=10,
                      min_samples_split=15, n_estimators=300, random_state=2)

In [151]:
rf_reg = RandomForestRegressor(max_depth=7, max_features=10, min_samples_leaf=10,
                      min_samples_split=15, n_estimators=300, random_state=2)

In [152]:
rf_reg.fit(X_train, y_train)

RandomForestRegressor(max_depth=7, max_features=10, min_samples_leaf=10,
                      min_samples_split=15, n_estimators=300, random_state=2)

In [153]:
y_hat = rf_reg.predict(X_test)

In [154]:
rmse = math.sqrt(mean_squared_error(y_test, y_hat))
rmse

1807.0486558756766

In [155]:
from sklearn import metrics
metrics.r2_score(y_test, y_hat)

0.8368392078688364

In [156]:
rf_reg = RandomForestRegressor(random_state=2)

In [157]:
rf_reg.fit(X_train, y_train)
y_hat = rf_reg.predict(X_test)
rmse = math.sqrt(mean_squared_error(y_test, y_hat))
print('RMSE: ',rmse)
r2 = metrics.r2_score(y_test, y_hat)
print('R-Square: ', r2)

RMSE:  1370.167296051605
R-Square:  0.9061955790277181


#### Best scores are obtained with the default values of the parameters in Random Forest
#### Paramter tuning didnt perform good
# RMSE:  1370.167296051605
# R-Square:  0.9061955790277181